include/deal.II/matrix_free/matrix_free.h

```diff
@@ -2081,257 +2081,192 @@ namespace internal
-      {};
-      tbb::task *execute ()
-      {
-        unsigned int lower = task_info.partition_color_blocks_data[partition],
-                     upper = task_info.partition_color_blocks_data[partition+1];
-        parallel_for(tbb::blocked_range<unsigned int>(lower,upper,1),
-                     CellWork<Worker> (worker,task_info));
-        if (is_blocked==true)
-          dummy->spawn (*dummy);
-        return (nullptr);
-      }
+      if (face_function != nullptr && face_range.second > face_range.first)
+        (container.*face_function)(matrix_free, this->dst, this->src, face_range);
+    }
 
-      tbb::empty_task *dummy;
+    // Runs the assembler on boundary faces. If no function is given, nothing
+    // is done
+    virtual void boundary(const std::pair<unsigned int,unsigned int> &face_range) override
+    {
+      if (boundary_function != nullptr && face_range.second > face_range.first)
+        (container.*boundary_function)(matrix_free, this->dst, this->src, face_range);
+    }
 
-    private:
-      const Worker &worker;
-      const unsigned int partition;
-      const internal::MatrixFreeFunctions::TaskInfo &task_info;
-      const bool is_blocked;
-    };
+    // Starts the communication for the update ghost values operation. We
+    // cannot call this update if ghost and destination are the same because
+    // that would introduce spurious entries in the destination (there is also
+    // the problem that reading from a vector that we also write to is usually
+    // not intended in case there is overlap, but this is up to the
+    // application code to decide and we cannot catch this case here).
+    virtual void vector_update_ghosts_start() override
+    {
+      if (!src_and_dst_are_same)
+        ghosts_were_set = internal::update_ghost_values_start(src);
+    }
 
-  } // end of namespace color
+    // Finishes the communication for the update ghost values operation
+    virtual void vector_update_ghosts_finish() override
+    {
+      if (!src_and_dst_are_same)
+        internal::update_ghost_values_finish(src);
+    }
 
+    // Starts the communication for the vector compress operation
+    virtual void vector_compress_start() override
+    {
+      internal::compress_start(dst);
+    }
 
-  template <typename VectorStruct>
-  class MPIComDistribute : public tbb::task
-  {
-  public:
-    MPIComDistribute (const VectorStruct  &src_in)
-      :
-      src(src_in)
-    {};
+    // Finishes the communication for the vector compress operation
+    virtual void vector_compress_finish() override
+    {
+      internal::compress_finish(dst);
+      if (!src_and_dst_are_same)
+        internal::reset_ghost_values(src, !ghosts_were_set);
+    }
 
-    tbb::task *execute ()
+    // Zeros the given input vector
+    virtual void zero_dst_vector_range(const unsigned int /*range_index*/) override
     {
-      internal::update_ghost_values_finish(src);
-      return nullptr;
+      // currently not implemented
```

Comment:
should not one add an Assert?

Reply:
This should arrive with a later commit in a few days.